src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,575 @@
+                else:
+                    lr = self.policy.distribution.prob(
+                        i.valid.action_var) / tf.stop_gradient(
+                            self.old_policy.distribution.prob(
+                                i.valid.action_var))
+
+                    surrogate = lr * adv_valid
+
+            # Finalize objective function
+            with tf.name_scope('loss'):
+                if self._pg_loss == 'vanilla':
+                    # VPG uses the vanilla objective
+                    obj = tf.identity(vanilla, name='vanilla_obj')
+                elif self._pg_loss == 'surrogate':
+                    # TRPO uses the standard surrogate objective
+                    obj = tf.identity(surrogate, name='surr_obj')
+                elif self._pg_loss == 'surrogate_clip':
+                    lr_clip = tf.clip_by_value(lr,
+                                               1 - self.lr_clip_range,
+                                               1 + self.lr_clip_range,
+                                               name='lr_clip')
+                    if self.policy.recurrent:
+                        surr_clip = lr_clip * adv * i.valid_var
+                    else:
+                        surr_clip = lr_clip * adv_valid
+                    obj = tf.minimum(surrogate, surr_clip, name='surr_obj')
+
+                if self._entropy_regularzied:
+                    obj += self.policy_ent_coeff * policy_entropy
+
+                # Maximize E[surrogate objective] by minimizing
+                # -E_t[surrogate objective]
+                if self.policy.recurrent:
+                    loss = -tf.reduce_sum(obj) / tf.reduce_sum(i.valid_var)
+                else:
+                    loss = -tf.reduce_mean(obj)
+
+            # Diagnostic functions
+            self.f_policy_kl = tensor_utils.compile_function(
+                flatten_inputs(self._policy_opt_inputs), pol_mean_kl)
+
+            self.f_rewards = tensor_utils.compile_function(
+                flatten_inputs(self._policy_opt_inputs), rewards)
+
+            returns = discounted_returns(self.discount, self.max_path_length,
+                                         rewards)
+            self.f_returns = tensor_utils.compile_function(
+                flatten_inputs(self._policy_opt_inputs),
+                returns,
+            )
+
+            return loss, pol_mean_kl
+
+    def _build_entropy_term(self, i):
+        pol_dist = self.policy.distribution
+        with tf.name_scope('policy_entropy'):
+            if self.policy.recurrent:
+                policy_neg_log_likeli = -pol_dist.log_prob(
+                    i.action_var, name='policy_log_likeli')
+
+                if self._use_neg_logli_entropy:
+                    policy_entropy = policy_neg_log_likeli
+                else:
+                    policy_entropy = pol_dist.entropy()
+            else:
+                policy_neg_log_likeli_flat = -pol_dist.log_prob(
```

Comment:
this is where i don't understand how your policy could possibly have the right shape for `pol_dist.log_prob(i.flat.action_var)` because `i.flat.action_var` needs to be shape `(batch * max_path_length)` so that it can be reshaped into `(batch, max_path_length)`. If you already flattened and filtered the observations, how could your policy be outputting the right shape?

Reply:
You are right, this will break. Looks like model should accept a placeholder instead of creating one, so we can pass a [batch, path, action_dim] placeholder and keep the original flattening / valid filtering logic. And we don't need the `model.input` as we are passing the placeholder.